src/firebolt/async_db/cursor.py

```diff
@@ -260,39 +235,105 @@ def _reset(self) -> None:
+        self._row_sets.append(row_set)
+        if self._next_set_idx == 0:
+            # Populate values for first set
+            self._pop_next_set()
+
+    async def _api_request(
+        self, query: str, set_parameters: Optional[dict]
+    ) -> Response:
+        return await self._client.request(
+            url="/",
+            method="POST",
+            params={
+                "database": self.connection.database,
+                "output_format": JSON_OUTPUT_FORMAT,
+                **self._set_parameters,
+                **(set_parameters or dict()),
+            },
+            content=query,
+        )
+
+    async def _do_execute(
+        self,
+        raw_query: str,
         parameters: Sequence[Sequence[ParameterType]],
         set_parameters: Optional[Dict] = None,
     ) -> None:
         self._reset()
+        if set_parameters is not None:
+            logger.warning(
+                "Passing set parameters as an argument is deprecated. Please run "
+                "a query 'SET <param> = <value>'"
+            )
         try:
 
-            queries = split_format_sql(query, parameters)
+            queries = split_format_sql(raw_query, parameters)
 
-            for query in queries:
+            for query in queries:  # type: ignore
 
                 start_time = time.time()
                 # our CREATE EXTERNAL TABLE queries currently require credentials,
                 # so we will skip logging those queries.
                 # https://docs.firebolt.io/sql-reference/commands/ddl-commands#create-external-table
-                if not re.search("aws_key_id|credentials", query, flags=re.IGNORECASE):
+                if isinstance(query, Set) or not re.search(
+                    "aws_key_id|credentials", query, flags=re.IGNORECASE
+                ):
                     logger.debug(f"Running query: {query}")
 
-                resp = await self._client.request(
-                    url="/",
-                    method="POST",
-                    params={
-                        "database": self.connection.database,
-                        "output_format": JSON_OUTPUT_FORMAT,
-                        **(set_parameters or dict()),
-                    },
-                    content=query,
-                )
+                # Define type for mypy
+                row_set: Tuple[
+                    int, Optional[List[Column]], Optional[List[List[RawColType]]]
+                ] = (-1, None, None)
+                if isinstance(query, Set):
+                    # Validate parameter by executing simple query with it
+                    resp = await self._api_request(
+                        "select 1", {query.name: query.value}
+                    )
+                    # Handle invalid set parameter
+                    if resp.status_code == codes.BAD_REQUEST:
+                        raise OperationalError(resp.text)
+                    await self._raise_if_error(resp)
```

Comment:
Wouldn't `_raise_if_error` be sufficient here? Why do we need L324-325?

Reply:
To return `OperationalError` with message in case of an invalid parameter. Inside of `_raise_if_error`, we call `raise_for_status`, which only returns error code, and is sufficient in most cases except for this one